File: lightroom_tagger/core/identity_service/percentiles.py
"""Within-perspective percentile ranks over the eligible identity population."""

from __future__ import annotations

import sqlite3
from collections import defaultdict
from typing import Any

from .aggregates import (
    _SCORES_BASE_SQL,
    _active_perspective_slugs,
    _default_min_perspectives,
    _truncate_rationale,
)


def _midrank_percentile_ranks(scores: list[int]) -> dict[int, float]:
    """Map each distinct raw score to a percentile rank in [0, 1] (midrank ties)."""
    n = len(scores)
    if n == 0:
        return {}
    if n == 1:
        return {scores[0]: 1.0}

    counts: dict[int, int] = defaultdict(int)
    for s in scores:
        counts[s] += 1

    out: dict[int, float] = {}
    below = 0
    for score in sorted(counts):
        tied = counts[score]
        # 1-based average rank for tied values.
        midrank = below + (tied + 1) / 2.0
        out[score] = (midrank - 1) / (n - 1)
        below += tied
    return out


def compute_within_perspective_percentile_lookup(
    conn: sqlite3.Connection,
) -> dict[tuple[str, str], float]:
    """Percentile rank in [0, 1] for each (image_key, perspective_slug) score cell.

  Computed once over the eligible population from :data:`_SCORES_BASE_SQL`
  (``is_current=1``, ``image_type='catalog'``, ``not_attempted=0``, active
  perspectives), with midrank tie handling within each perspective.
    """
    rows = conn.execute(_SCORES_BASE_SQL).fetchall()
    scores_by_perspective: dict[str, list[int]] = defaultdict(list)
    cells: list[tuple[str, str, int]] = []

    for r in rows:
        image_key = str(r["image_key"])
        slug = str(r["perspective_slug"])
        score = int(r["score"])
        scores_by_perspective[slug].append(score)
        cells.append((image_key, slug, score))

    score_percentile_by_perspective = {
        slug: _midrank_percentile_ranks(scores)
        for slug, scores in scores_by_perspective.items()
    }

    lookup: dict[tuple[str, str], float] = {}
    for image_key, slug, score in cells:
        lookup[(image_key, slug)] = score_percentile_by_perspective[slug][score]
    return lookup


def compute_image_peak_percentile_scores(
    conn: sqlite3.Connection,
    *,
    min_perspectives: int | None = None,
    include_ineligible: bool = True,
) -> tuple[list[dict[str, Any]], dict[str, Any]]:
    """Per-image peak within-perspective percentile plus per-perspective detail.

    Returns ``(items, meta)``. Each item includes ``peak_percentile`` (max
    percentile across the image's scored lenses), ``perspectives_covered``,
  ``eligible``, and ``per_perspective`` entries with ``percentile``.
    """
    active_slugs = _active_perspective_slugs(conn)
    active_count = len(active_slugs)
    slug_set = set(active_slugs)
    min_used = (
        int(min_perspectives)
        if min_perspectives is not None
        else _default_min_perspectives(active_count)
    )

    total_catalog = int(
        conn.execute("SELECT COUNT(*) AS c FROM images").fetchone()["c"]
    )

    percentile_lookup = compute_within_perspective_percentile_lookup(conn)
    rows = conn.execute(_SCORES_BASE_SQL).fetchall()

    by_key: dict[str, list[dict[str, Any]]] = {}
    for r in rows:
        slug = str(r["perspective_slug"])
        if slug not in slug_set:
            continue
        image_key = str(r["image_key"])
        score = int(r["score"])
        percentile = percentile_lookup[(image_key, slug)]
        by_key.setdefault(image_key, []).append(
            {
                "perspective_slug": slug,
                "display_name": r["perspective_display_name"] or slug,
                "score": score,
                "percentile": round(percentile, 6),
                "rationale": r.get("rationale") or "",
                "model_used": r.get("model_used") or "",
                "prompt_version": r.get("prompt_version") or "",
                "scored_at": r.get("scored_at") or "",
            }
        )

    items: list[dict[str, Any]] = []
    eligible_count = 0
    for image_key, perspectives in by_key.items():
        n = len(perspectives)
        peak = max(p["percentile"] for p in perspectives) if perspectives else 0.0
        eligible = n >= min_used
        if eligible:
            eligible_count += 1

        per_out: list[dict[str, Any]] = []
        for p in sorted(perspectives, key=lambda x: x["perspective_slug"]):
            per_out.append(
                {
                    "perspective_slug": p["perspective_slug"],
                    "display_name": p["display_name"],
                    "score": p["score"],
                    "percentile": p["percentile"],
                    "prompt_version": p["prompt_version"],
                    "model_used": p["model_used"],
                    "scored_at": p["scored_at"],
                    "rationale_preview": _truncate_rationale(p.get("rationale")),
                }
            )

        row = {
            "image_key": image_key,
            "peak_percentile": round(peak, 6),
            "perspectives_covered": n,
            "eligible": eligible,
            "per_perspective": per_out,
        }
        if include_ineligible or eligible:
            items.append(row)

    scored_any_count = len(by_key)
    coverage_rule = "eligible when perspectives_covered >= min_perspectives (default 1)"
    meta: dict[str, Any] = {
        "active_perspectives": active_slugs,
        "weighting": "peak_within_perspective_percentile",
        "ranking_key": "peak_percentile",
        "min_perspectives_used": min_used,
        "coverage_rule": coverage_rule,
        "total_catalog_images": total_catalog,
        "eligible_count": eligible_count,
        "scored_any_count": scored_any_count,
    }
    if eligible_count == 0 and active_count > 0:
        meta["coverage_note"] = (
            "No images meet the minimum perspective coverage for ranking; "
            "score at least one perspective per image."
        )
    return items, meta

File: apps/visualizer/backend/src/identity/percentiles.ts
/**
 * Within-perspective percentile ranks over the eligible identity population.
 *
 * A raw 1–10 score is not comparable across perspectives: one lens may hand out
 * 8s freely while another rarely exceeds 5. Ranking on the *percentile within each
 * perspective* is what makes "this photo's strongest lens" mean something.
 */
import type { Db } from '../db/connection.js';
import {
  activePerspectiveSlugs,
  defaultMinPerspectives,
  SCORES_BASE_SQL,
  truncateRationale,
  type ScoreRow,
} from './aggregates.js';

/**
 * Round to 6 decimals, matching `round(value, 6)`.
 *
 * Python rounds halves to even and JavaScript rounds them away from zero. Landing
 * exactly on a half at the seventh decimal requires the percentile — a ratio of the
 * form `(midrank - 1) / (n - 1)` — to terminate at exactly seven decimal places,
 * which needs `n - 1` to be a product of 2s and 5s and the numerator to be odd in
 * just the right way. The real-catalog parity test compares every published
 * percentile against Python's, so this is verified rather than argued.
 */
export function round6(value: number): number {
  return Math.round(value * 1e6) / 1e6;
}

/**
 * Map each distinct raw score to a percentile rank in [0, 1], with midrank ties.
 *
 * Midrank (rather than "fraction below") means a tied group all get the average of
 * the ranks they span, so a lens that gives half the catalog a 7 does not make
 * every one of those photos look either top or bottom of the pile.
 */
export function midrankPercentileRanks(scores: readonly number[]): Map<number, number> {
  const n = scores.length;
  if (n === 0) return new Map();
  if (n === 1) return new Map([[scores[0]!, 1]]);

  const counts = new Map<number, number>();
  for (const s of scores) counts.set(s, (counts.get(s) ?? 0) + 1);

  const out = new Map<number, number>();
  let below = 0;
  // Numeric sort: these are score values, and the default string sort would put
  // 10 before 2.
  for (const score of [...counts.keys()].sort((a, b) => a - b)) {
    const tied = counts.get(score)!;
    // 1-based average rank across the tied group.
    const midrank = below + (tied + 1) / 2;
    out.set(score, (midrank - 1) / (n - 1));
    below += tied;
  }
  return out;
}

/**
 * Compose the `(image_key, perspective_slug)` lookup key.
 *
 * The separator is an explicit NUL escape rather than a literal character: no image
 * key or slug can contain one, so the composite key is unambiguous, and writing it
 * as `\u0000` keeps it visible in the source. Exported so every caller composes the
 * key through this one function — building it inline in a second module is how a
 * mismatched separator silently turns every percentile into NaN.
 */
export function cellKey(imageKey: string, slug: string): string {
  return `${imageKey}\u0000${slug}`;
}

/**
 * Percentile rank in [0, 1] for every `(image_key, perspective_slug)` score cell.
 *
 * Computed once over the whole eligible population, because a percentile is
 * meaningless relative to a page of results.
 */
export function computeWithinPerspectivePercentileLookup(
  db: Db,
  preloadedRows?: readonly ScoreRow[],
): Map<string, number> {
  // Callers that already hold the score rows pass them in; on this catalog the
  // query returns 146,983 rows, so reading it twice is not free.
  const rows = preloadedRows ?? (db.prepare(SCORES_BASE_SQL).all() as ScoreRow[]);

  const scoresByPerspective = new Map<string, number[]>();
  const cells: [string, string, number][] = [];
  for (const r of rows) {
    const imageKey = String(r.image_key);
    const slug = String(r.perspective_slug);
    const score = Math.trunc(r.score);
    const list = scoresByPerspective.get(slug);
    if (list) list.push(score);
    else scoresByPerspective.set(slug, [score]);
    cells.push([imageKey, slug, score]);
  }

  const rankBySlug = new Map<string, Map<number, number>>();
  for (const [slug, scores] of scoresByPerspective) {
    rankBySlug.set(slug, midrankPercentileRanks(scores));
  }

  const lookup = new Map<string, number>();
  for (const [imageKey, slug, score] of cells) {
    lookup.set(cellKey(imageKey, slug), rankBySlug.get(slug)!.get(score)!);
  }
  return lookup;
}

export const CORROBORATION_RULE =
  'If any lens scored 1, rank on second-highest percentile instead of peak.';

export interface PerspectiveCell {
  perspective_slug: string;
  display_name: string;
  score: number;
  percentile: number;
  rationale: string;
  model_used: string;
  prompt_version: string;
  scored_at: string;
}

export interface CorroborationFields {
  ranking_percentile: number;
  corroboration_revoked: boolean;
  corroboration_revoked_by: string;
}

/**
 * The corroboration veto (#292): one flattering lens cannot crown a photo.
 *
 * If any lens scored the image a 1, ranking falls back from the peak percentile to
 * the *second*-highest. The intent is that a photo another lens considers a total
 * failure should not top the list on one enthusiastic reading.
 */
export function corroborationRankingFields(
  perspectives: readonly PerspectiveCell[],
  peak: number,
): CorroborationFields {
  const unrevoked = (): CorroborationFields => ({
    ranking_percentile: round6(peak),
    corroboration_revoked: false,
    corroboration_revoked_by: '',
  });

  if (perspectives.length === 0) return unrevoked();

  const minScore = Math.min(...perspectives.map((p) => Math.trunc(p.score)));
  if (minScore > 1) return unrevoked();

  const sortedPercentiles = perspectives.map((p) => Number(p.percentile)).sort((a, b) => b - a);
  // A single-lens image that scored 1 has no second opinion to fall back to, so it
  // ranks at the bottom rather than on its own peak.
  const ranking = sortedPercentiles.length >= 2 ? sortedPercentiles[1]! : 0;

  // Lowest score first, then slug, so the reported culprit is deterministic.
  const revoking = perspectives
    .filter((p) => Math.trunc(p.score) <= 1)
    .sort((a, b) => {
      const byScore = Math.trunc(a.score) - Math.trunc(b.score);
      if (byScore !== 0) return byScore;
      return a.perspective_slug < b.perspective_slug ? -1 : a.perspective_slug > b.perspective_slug ? 1 : 0;
    });

  return {
    ranking_percentile: round6(ranking),
    corroboration_revoked: true,
    corroboration_revoked_by: revoking.length > 0 ? String(revoking[0]!.perspective_slug) : '',
  };
}

export interface PeakPercentileItem extends CorroborationFields {
  image_key: string;
  peak_percentile: number;
  perspectives_covered: number;
  eligible: boolean;
  per_perspective: {
    perspective_slug: string;
    display_name: string;
    score: number;
    percentile: number;
    prompt_version: string;
    model_used: string;
    scored_at: string;
    rationale_preview: string;
  }[];
}

export interface PeakPercentileMeta {
  active_perspectives: string[];
  weighting: string;
  ranking_key: string;
  corroboration_rule: string;
  min_perspectives_used: number;
  coverage_rule: string;
  total_catalog_images: number;
  eligible_count: number;
  scored_any_count: number;
  coverage_note?: string;
}

/**
 * Per-image peak within-perspective percentile plus per-perspective detail.
 *
 * `percentileLookup` can be passed in when the caller already built one — the
 * Mirror does, and recomputing it would mean a second full pass over every score in
 * the catalog.
 */
export function computeImagePeakPercentileScores(
  db: Db,
  opts: {
    minPerspectives?: number | null;
    includeIneligible?: boolean;
    percentileLookup?: Map<string, number> | null;
  } = {},
): { items: PeakPercentileItem[]; meta: PeakPercentileMeta } {
  const includeIneligible = opts.includeIneligible ?? true;
  const activeSlugs = activePerspectiveSlugs(db);
  const activeCount = activeSlugs.length;
  const slugSet = new Set(activeSlugs);
  const minUsed =
    opts.minPerspectives !== null && opts.minPerspectives !== undefined
      ? Math.trunc(opts.minPerspectives)
      : defaultMinPerspectives(activeCount);

  const totalCatalog = Math.trunc(
    (db.prepare('SELECT COUNT(*) AS c FROM images').get() as { c: number }).c,
  );

  const percentileLookup =
    opts.percentileLookup ?? computeWithinPerspectivePercentileLookup(db);
  const rows = db.prepare(SCORES_BASE_SQL).all() as ScoreRow[];

  // Insertion order matters: it becomes the pre-sort order of `items`, and the
  // callers' sorts are stable, so it is the final tiebreaker.
  const byKey = new Map<string, PerspectiveCell[]>();
  for (const r of rows) {
    const slug = String(r.perspective_slug);
    if (!slugSet.has(slug)) continue;
    const imageKey = String(r.image_key);
    const score = Math.trunc(r.score);
    const percentile = percentileLookup.get(cellKey(imageKey, slug))!;
    const cell: PerspectiveCell = {
      perspective_slug: slug,
      display_name: r.perspective_display_name || slug,
      score,
      percentile: round6(percentile),
      rationale: r.rationale ?? '',
      model_used: r.model_used ?? '',
      prompt_version: r.prompt_version ?? '',
      scored_at: r.scored_at ?? '',
    };
    const list = byKey.get(imageKey);
    if (list) list.push(cell);
    else byKey.set(imageKey, [cell]);
  }

  const items: PeakPercentileItem[] = [];
  let eligibleCount = 0;
  for (const [imageKey, perspectives] of byKey) {
    const n = perspectives.length;
    const peak = n > 0 ? Math.max(...perspectives.map((p) => p.percentile)) : 0;
    const vetoFields = corroborationRankingFields(perspectives, peak);
    const eligible = n >= minUsed;
    if (eligible) eligibleCount += 1;

    const perOut = [...perspectives]
      .sort((a, b) =>
        a.perspective_slug < b.perspective_slug ? -1 : a.perspective_slug > b.perspective_slug ? 1 : 0,
      )
      .map((p) => ({
        perspective_slug: p.perspective_slug,
        display_name: p.display_name,
        score: p.score,
        percentile: p.percentile,
        prompt_version: p.prompt_version,
        model_used: p.model_used,
        scored_at: p.scored_at,
        rationale_preview: truncateRationale(p.rationale),
      }));

    if (includeIneligible || eligible) {
      items.push({
        image_key: imageKey,
        peak_percentile: round6(peak),
        ...vetoFields,
        perspectives_covered: n,
        eligible,
        per_perspective: perOut,
      });
    }
  }

  const meta: PeakPercentileMeta = {
    active_perspectives: activeSlugs,
    weighting: 'peak_within_perspective_percentile',
    ranking_key: 'ranking_percentile',
    corroboration_rule: CORROBORATION_RULE,
    min_perspectives_used: minUsed,
    coverage_rule: 'eligible when perspectives_covered >= min_perspectives (default 1)',
    total_catalog_images: totalCatalog,
    eligible_count: eligibleCount,
    scored_any_count: byKey.size,
  };
  if (eligibleCount === 0 && activeCount > 0) {
    meta.coverage_note =
      'No images meet the minimum perspective coverage for ranking; ' +
      'score at least one perspective per image.';
  }
  return { items, meta };
}
